bpo-43605: Improve the documentation to exec() and eval()

congma opened the pull request:
- Add links to the relevant section in Language Reference about dynamic
  execution's limitation with respect to namespaces.
- For eval(), move some explanatory text into a "Note" box.
- Make the first paragraph of eval's doc consistent about what the first
  argument accepts.
- For exec(), remove the text about the "globals" optional argument
  having to be a dict but not an instance of a subclass of dict. This is
  no longer true -- the code calls PyDict_Check(), not
  PyDict_CheckExact().
- Put quotes around the ``__builtins__`` in the text: clarify that in
  the context it means a string key in the dict passed to eval/exec as
  the globals dict. Otherwise, since the identifier __builtins__ refers
  to a module, it can be confusing and misleading.
- Reordering some paragraphs so that overall the structure is improved.
- Re-wrap some long lines in RST source.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-43605](https://bugs.python.org/issue43605) -->
https://bugs.python.org/issue43605
<!-- /issue-number -->

congma committed:
bpo-43605: Improve the documentation to exec() and eval()

- Add links to the relevant section in Language Reference about dynamic
  execution's limitation with respect to namespaces.
- For eval(), move some explanatory text into a "Note" box.
- Make the first paragraph of eval's doc consistent about what the first
  argument accepts.
- For exec(), remove the text about the "globals" optional argument
  having to be a dict but not an instance of a subclass of dict. This is
  no longer true -- the code calls PyDict_Check(), not
  PyDict_CheckExact().
- Put quotes around the ``__builtins__`` in the text: clarify that in
  the context it means a string key in the dict passed to eval/exec as
  the globals dict. Otherwise, since the identifier __builtins__ refers
  to a module, it can be confusing and misleading.
- Reordering some paragraphs so that overall the structure is improved.
- Re-wrap some long lines in RST source.

Fidget-Spinner commented on Doc/library/functions.rst:
```diff
@@ -496,44 +496,38 @@ are always available.  They are listed here in alphabetical order.
-
-   The return value is the result of
-   the evaluated expression. Syntax errors are reported as exceptions.  Example:
+   This function supports the dynamic evaluation of Python expression.  The
```
```suggestion
   This function supports the dynamic evaluation of Python expressions.  The
```

congma replied:
Good catch! Thank you!

Fidget-Spinner commented on Doc/library/functions.rst:
```diff
@@ -496,44 +496,38 @@ are always available.  They are listed here in alphabetical order.
+   This function supports the dynamic evaluation of Python expression.  The
+   first argument can be a string or a code object.  The optional arguments
+   specify the globals and locals respectively.  If provided, *globals* must be
+   a dictionary.  If provided, *locals* can be any mapping object.
```
I don't think `locals` can be any mapping object. Eg. ``collections.Counter()`` is a mapping object (``isinstance(collections.Counter('a'), collections.abc.Mapping)`` is ``True``), but it cannot be used for eval:
```python
>>> Counter("abcd")
Counter({'a': 1, 'b': 1, 'c': 1, 'd': 1})
>>> eval("print(a)", {}, Counter("abcd"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
TypeError: 'int' object is not callable
>>> eval("print(a)", {}, {'a':1})
1
```


Surprisingly, git blame says this line has been around since this paragraph's inception 14 years ago.

```suggestion
   specify the globals and locals respectively.  If provided, *globals*
   and *locals* must be dictionaries.
```

congma replied:
What?? In the source for `eval/exec` in Python/bltinmodule.c, the locals are checked by `PyMapping_Check()` while globals checked by `PyDict_Check()`. I'm not sure why the exception was about "'int' object is not callable" -- it seems the name `a` did get resolved in the locals through the `Counter` object, but somehow there was an action to the effect of `1()`. <s>This might be a bug! Right now I don't have enough spoons to track it down, but I should be verifying it and maybe submitting a bug report soon.</s> Edit: working as intended, but in a tricky way: see [my reply below](https://github.com/python/cpython/pull/25039#discussion_r602783224).

congma replied:
This works as intended:
```python
>>> ct = Counter("abcd")
>>> eval("a", {}, ct)
1
```
But not this:
```python
>>> eval("print(a)", {}, ct)
```
Expected return value of `eval` is `None` (return value of `print` function), but the cryptic exception is raised.

congma replied:
Also, this:
```python
>>> eval("print(1)", {}, ct)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
TypeError: 'int' object is not callable
```

congma replied:
I see! The locals have the highest priority in name resolution. The `Counter` instance (`ct` in the above example) behaves as if any key is in it (with initial value 0). When the machinery underlying `eval` does its work, it first looks up the name `"print"` in the locals given to it. Unfortunately, as a result of how `Counter` behaves, it gets what it wishes for: the count for the key `"print"` in the counter `ct`, which is the `int` instance `0`. Then `0` gets called as the function.  The evaluation boils down to `0(1)` which raises `TypeError`.

Fidget-Spinner replied:
Awesome! Thank you for investigating this! I learnt a lot :).

I forgot that Counter sort of behaves like a defaultdict and returns 0 for keys it doesn't have.

congma replied:
Thanks, me too. This is the weirdest thing I've seen in Python in a while.

Fidget-Spinner commented on Doc/library/functions.rst:
```diff
@@ -496,44 +496,38 @@ are always available.  They are listed here in alphabetical order.
+
+   In the most common case, the *expression* argument is a string, and it is
+   parsed and evaluated as a Python expression (see the section
+   ":ref:`expression-input`" in the Language Reference).  The leading and
```
Sorry, I don't remember if this is valid/invalid inline markup. Some roles require whitespace around them. See this to learn more:
https://devguide.python.org/documenting/#inline-markup

So it might be ``"\ :ref:`expression-input`\ "``. However, if you ran ``./make html`` and all looks fine on your computer, then I'm wrong and please ignore this comment :).

congma replied:
Hi, I think this works fine (verified with the Sphinx setup created by "make venv" and "make html") -- unlike inline markup, the `ref` cross-linking markup seem to be Ok near punctuations.

Fidget-Spinner reviewed (commented):
Thanks for submitting this PR to improve Python docs! I think this looks good overall, though I have a few minor comments below.

Additionally, I wonder if this would've been better if it were split into multiple smaller PRs? I found this slightly long to review.

Fidget-Spinner reviewed (approved):
LGTM. I think this might need a news item (sometimes significant docs changes require them). Lets see what the core dev says when they review this.

congma committed:
Further revise eval/exec documentations for consistency and accuracy.

The documentations for eval/exec are further modified for consistency
and accuracy.

- For both functions, the documentation mostly follows the following
  structure: description of arguments, default values for optional
  arguments, special treatment of the "__builtins__" key, and notices
  about limitations or other noteworthy material. Previously, the
  description about the special "__builtins__" key is intermingled with
  the description about default values for missing arguments. Now, these
  are in separate paragraphs. Also, it is stressed that the special key
  is inserted in-place.
- In the notice about limitations to dynamic eval/exec w.r.t. scopes,
  also mention the assignment expression succinctly.
- Remove the description that exec() with both globals and locals behave
  "as if" the source string is embedded in a class block. Again, a class
  block inside a function can access the non-locals, but this is not
  available to exec().
- Instead of mentioning passing the return values of globals() and
  locals() around, we now say that the programmer can construct their
  own namespaces based on their return values. The reason is that
  currently the semantics of locals() is not very well defined, and in
  the light of PEP 558 it may change in the future.  Explicitly
  constructing new namespaces is more predictable than using the "raw"
  return value of locals(). In the future, this part of the docs can be
  further updated to keep up with the new definition of locals().
- Omit the footnote to exec() about the parser only accepting Unix-style
  LF newlines. This piece of information is outdated.
- Other small fixes for grammar.

congma committed:
Fix invalid RST syntax.

JelleZijlstra commented on Doc/library/functions.rst:
```diff
@@ -496,44 +496,41 @@ are always available.  They are listed here in alphabetical order.
-
-   The return value is the result of
-   the evaluated expression. Syntax errors are reported as exceptions.  Example:
+   This function supports the dynamic evaluation of Python expressions.  The
```
```suggestion
   This function dynamically evaluates Python expressions.  The
```

JelleZijlstra commented on Doc/library/functions.rst:
```diff
@@ -496,44 +496,41 @@ are always available.  They are listed here in alphabetical order.
-   The return value is the result of
-   the evaluated expression. Syntax errors are reported as exceptions.  Example:
+   This function supports the dynamic evaluation of Python expressions.  The
+   first argument can be a string or a code object.  The optional arguments
```
Maybe link this to the documentation for code objects.

JelleZijlstra commented on Doc/library/functions.rst:
```diff
@@ -496,44 +496,41 @@ are always available.  They are listed here in alphabetical order.
+   *globals* must be a dictionary.  If provided, *locals* can be any
+   :term:`mapping` object.
+
+   In the most common case, the *expression* argument is a string, and it is
```
```suggestion
   When the *expression* argument is a string, it is
```
Tighten the wording

JelleZijlstra commented on Doc/library/functions.rst:
```diff
@@ -496,44 +496,41 @@ are always available.  They are listed here in alphabetical order.
+   given, the local namespace defaults to *globals*.
+
+   Before evaluation, the special key ``"__builtins__"`` is searched for in the
+   global namespace that is explicitly or impcitly specified for :func:`eval`.
```
```suggestion
   global namespace that is explicitly or implicitly specified.
```

JelleZijlstra commented on Doc/library/functions.rst:
```diff
@@ -543,36 +540,60 @@ are always available.  They are listed here in alphabetical order.
+      :ref:`comprehensions <comprehensions>`, and :term:`generator expressions
+      <generator expression>` which create an inner scope of their own.  Since
+      Python 3.8, the action of an assignment expression (see :pep:`572`) also
+      depends on scope information determined at compile-time.  Interaction
```
```suggestion
      depends on scope information determined at compile time.  Interaction
```

iritkatriel reviewed (changes requested):
This has merge conflicts now.
